fetcher/fetcher_test.go

```diff
@@ -0,0 +1,123 @@
+	aclient := mockAClient(t, &AlgodHandler{})
+	fetcher := &fetcherImpl{aclient: aclient, log: logrus.New()}
+	require.Equal(t, aclient, fetcher.Algod(), "Algod client returned from fetcherImpl does not match expected instance.")
+}
+
+func TestFetcherImplSetError(t *testing.T) {
+	aclient := mockAClient(t, &AlgodHandler{})
+	fetcher := &fetcherImpl{aclient: aclient, log: logrus.New()}
+	expectedErr := fmt.Errorf("foobar")
+	fetcher.setError(expectedErr)
+	require.Equal(t, expectedErr.Error(), fetcher.Error(), "Error produced by setError was not reflected in Error output.")
+}
+
+func TestFetcherImplProcessQueueHandlerError(t *testing.T) {
+	mockAlgodHandler := &AlgodHandler{}
+	aclient := mockAClient(t, mockAlgodHandler)
+	ctx, cf := context.WithCancel(context.Background())
+	defer cf()
+	fetcher := &fetcherImpl{aclient: aclient, log: logrus.New()}
+	bHandler := &BlockHandler{}
+	expectedError := fmt.Errorf("handlerError")
+	// The block handler function will immediately return an error on any block passed to it
+	bHandler.On("handlerFunc", mock.Anything, mock.Anything).Return(expectedError)
+	fetcher.SetBlockHandler(bHandler.handlerFunc)
+	// Mock algod server to continually return empty blocks
+	mockAlgodHandler.On("ServeHTTP", mock.Anything, mock.Anything).Run(func(args mock.Arguments) {
+		respWriter := args.Get(0).(http.ResponseWriter)
+		req := args.Get(1).(*http.Request)
+		path := req.URL.Path
+		if strings.Contains(path, "v2/blocks/") {
+			var block = new(bookkeeping.Block)
+			respWriter.Write(protocol.Encode(block))
+		}
+	})
+	require.ErrorIsf(t, fetcher.Run(ctx), expectedError, "FetcherImpl did not return expected error in processQueue handler.")
+}
+
+func TestFetcherImplCatchupLoopBlockError(t *testing.T) {
+	mockAlgodHandler := &AlgodHandler{}
+	aclient := mockAClient(t, mockAlgodHandler)
+	ctx, cf := context.WithCancel(context.Background())
+	defer cf()
+	passingCalls := 5
+	// Initializing blockQueue here needs buffer since we have no other goroutines receiving from it
+	fetcher := &fetcherImpl{aclient: aclient, log: logrus.New(), blockQueue: make(chan *rpcs.EncodedBlockCert, 256)}
+	bHandler := &BlockHandler{}
+	// the handler will do nothing here
+	bHandler.On("handlerFunc", mock.Anything, mock.Anything).Return(nil)
+	fetcher.SetBlockHandler(bHandler.handlerFunc)
+
+	// Our mock algod client will process /v2/blocks/{round} calls
+	// returning an empty block `passingCalls` times before throwing 500s
+	mockAlgodHandler.On("ServeHTTP", mock.Anything, mock.Anything).Run(func(args mock.Arguments) {
+		respWriter := args.Get(0).(http.ResponseWriter)
+		req := args.Get(1).(*http.Request)
+		path := req.URL.Path
+		if strings.Contains(path, "v2/blocks/") {
+			if passingCalls <= 0 {
+				respWriter.WriteHeader(http.StatusInternalServerError)
+			} else {
+				block := new(bookkeeping.Block)
+				respWriter.WriteHeader(http.StatusOK)
+				respWriter.Write(protocol.Encode(block))
+				passingCalls--
```

Comment:
nit: I think it's pretty uncommon for `new` to be the right way to declare something. It's all personal preferences, but I would probably opt for one of these:
```suggestion
				respWriter.WriteHeader(http.StatusOK)
				respWriter.Write(protocol.Encode(&bookkeeping.Block{}))
				passingCalls--
```
```suggestion
				var block bookkeeping.Block
				respWriter.WriteHeader(http.StatusOK)
				respWriter.Write(protocol.Encode(&block))
				passingCalls--
```
```suggestion
				data := protocol.Encode(&bookkeeping.Block{})
				respWriter.WriteHeader(http.StatusOK)
				respWriter.Write(data)
				passingCalls--
```